server/auth.js

```diff
@@ -14,26 +15,71 @@ const dayjs = require("dayjs");
  * @returns {Promise<(Bean|null)>} User or null if login failed
  */
 exports.login = async function (username, password) {
+
     if (typeof username !== "string" || typeof password !== "string") {
         return null;
     }
-
-    let user = await R.findOne("user", " username = ? AND active = 1 ", [
-        username,
-    ]);
-
-    if (user && passwordHash.verify(password, user.password)) {
-        // Upgrade the hash to bcrypt
-        if (passwordHash.needRehash(user.password)) {
-            await R.exec("UPDATE `user` SET password = ? WHERE id = ? ", [
-                passwordHash.generate(password),
-                user.id,
-            ]);
+    if (process.env.AUTH_METHOD === "LDAP") {
+        let ldapAuthSuccess = false;
+        await new Promise((resolve, reject) => {
+            const client = ldap.createClient({
+                url: [ process.env.LDAP_ADDRESS ]
+            });
+            client.on("connectError", (err) => {
+                log.error("auth", "connecting ldap server failed");
+                resolve();
+            });
+            client.bind(process.env.LDAP_UID + "=" + username + "," + process.env.LDAP_BASE_DN,
+                password, (err) => {
+                    if (err) {
+                        log.warn("auth", "ldap authentication failed for user: " + username);
+                    } else {
+                        log.info("auth", "ldap authentication succeeded for user: " + username);
+                        ldapAuthSuccess = true;
+                    }
+                    resolve();
+                });
+        });
+        if (!ldapAuthSuccess) {
+            return null;
+        }
+        let user = await R.findOne("user", " username = ? AND active = 1 ", [
```

Comment:
Please extract the below parts into functions.
This is hard to comprehend => easy to make mistakes. Lets not risk this in security critical code ^^